src/Plugin/resource/DataProvider/DataProviderFile.php

```diff
@@ -0,0 +1,327 @@
+<?php
+
+/**
+ * @file
+ * Contains \Drupal\restful\Plugin\resource\DataProvider\DataProviderFile.
+ */
+
+namespace Drupal\restful\Plugin\resource\DataProvider;
+
+use Drupal\restful\Exception\BadRequestException;
+use Drupal\restful\Exception\NotImplementedException;
+use Drupal\restful\Exception\ServiceUnavailableException;
+use Drupal\restful\Http\RequestInterface;
+use Drupal\restful\Plugin\resource\Field\ResourceFieldCollectionInterface;
+
+class DataProviderFile extends DataProvider implements DataProviderInterface {
+
+  /**
+   * Constructs a DataProviderFile object.
+   */
+  public function __construct(RequestInterface $request, ResourceFieldCollectionInterface $field_definitions, $account, array $options, $langcode = NULL) {
+    parent::__construct($request, $field_definitions, $account, $options, $langcode);
+
+    $file_options = empty($this->options['options']) ? array() : $this->options['options'];
+    $default_values = array(
+      'validators' => array(
+        'file_validate_extensions' => array(),
+        'file_validate_size' => array(),
+      ),
+      'scheme' => file_default_scheme(),
+      'replace' => FILE_EXISTS_RENAME,
+    );
+    $this->options['options'] = drupal_array_merge_deep($default_values, $file_options);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function index() {
+    throw new NotImplementedException(sprintf('The %s method is not supported.', __FUNCTION__));
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function create($object) {
+    $files = $this->getRequest()->getFiles();
+    if (!$files) {
+      throw new BadRequestException('No files sent with the request.');
+    }
+
+    $ids = array();
+
+    foreach ($files as $file_info) {
+      // Populate the $_FILES the way file_save_upload() expects.
+      $name = $file_info['name'];
+      foreach ($file_info as $key => $value) {
+        $files['files'][$key][$name] = $value;
+      }
+
+      $file = $this->fileSaveUpload($name, $files);
+
+      // Change the file status from temporary to permanent.
+      $file->status = FILE_STATUS_PERMANENT;
+      file_save($file);
+
+      // Required to be able to reference this file.
+      file_usage_add($file, 'restful', 'files', $file->fid);
+
+      $ids[] = $file->fid;
+    }
+
+    $return = array();
+    foreach ($ids as $id) {
+      $return[] = $this->view($id);
+    }
+
+    return $return;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function view($identifier) {
+    // TODO: Provide real support for files. Reading files is probably a good example for the DB Data provider.
+    return array('fid' => $identifier);
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function viewMultiple(array $identifiers) {
```

Comment:
A file is an entity by default . File entity module just makes it fieldable

‫ב-4 באפר 2015, בשעה 09:42, ‏Mateu Aguiló Bosch <notifications@github.com> כתב/ה:‬

> In src/Plugin/resource/DataProvider/DataProviderFile.php:
> 
> > +
> > +    return $return;
> > +  }
> > +
> > +  /**
> > +   * {@inheritdoc}
> > +   */
> > +  public function view($identifier) {
> > +    // TODO: Provide real support for files. Reading files is probably a good example for the DB Data provider.
> > +    return array('fid' => $identifier);
> > +  }
> > +
> > +  /**
> > +   * {@inheritdoc}
> > +   */
> > +  public function viewMultiple(array $identifiers) {
> I did not want to depend on the file_entity module only for this.
> 
> —
> Reply to this email directly or view it on GitHub.
>